File: human-protocol-btc/kotlin/protocol/src/main/kotlin/org/hpb/protocol/Validators.kt
package org.hpb.protocol

import org.hpb.engine.hex
import org.hpb.engine.sha256

/**
 * Mechanical validation: pure functions over the revealed submission set.
 * Any observer recomputes both acceptance and the payout list — validation
 * needs no judge, only reveal.
 */
object Validators {
    fun normalize(answer: String): String = answer.trim().lowercase()

    /** The commitment format for groundtruth sets: sha256("key:normalized"). */
    fun groundtruthHash(taskKey: String, answer: String): String =
        sha256("$taskKey:${normalize(answer)}".toByteArray()).hex()

    /** (taskKey, worker) -> answer, from all revealed submissions. */
    data class Submitted(val taskKey: String, val worker: String, val answer: String)

    private val GRANTED = setOf(
        AssignmentStatus.ACTIVE, AssignmentStatus.SUBMITTED, AssignmentStatus.VALIDATED,
    )

    /**
     * Grant-scoping, applied by the launcher when collecting AND re-checked
     * by every witness against the reveal: a row counts only when its worker
     * holds a granted assignment covering the task, and only once per
     * (worker, task) — first occurrence wins. Without this, one worker
     * repeating an answer earns one reward per repetition.
     */
    fun scoped(rows: List<Submitted>, assignments: List<AssignmentState>): List<Submitted> {
        val granted = assignments.filter { it.status in GRANTED }
            .associate { it.worker to it.taskKeys.toSet() }
        val seen = HashSet<Pair<String, String>>()
        return rows.filter { row ->
            row.taskKey in granted[row.worker].orEmpty() && seen.add(row.worker to row.taskKey)
        }
    }

    fun validate(policy: ValidationPolicy, submissions: List<Submitted>): List<ResultRow> =
        when (policy.type) {
            ValidationType.GROUNDTRUTH -> submissions.map {
                ResultRow(
                    it.taskKey, it.worker, it.answer,
                    accepted = groundtruthHash(it.taskKey, it.answer) in policy.groundtruthHashes,
                )
            }
            ValidationType.AGREEMENT -> agreement(policy, submissions)
        }

    /**
     * Inter-worker consensus: per task, the modal normalized answer wins when
     * it reaches ceil(n * threshold) of that task's submissions; workers whose
     * answers match the winning cluster are accepted. Deterministic tie-break:
     * lexicographically smallest modal answer.
     */
    private fun agreement(
        policy: ValidationPolicy,
        submissions: List<Submitted>,
    ): List<ResultRow> {
        val winners = submissions.groupBy { it.taskKey }.mapValues { (_, subs) ->
            winningAnswer(policy, subs)
        }
        return submissions.map {
            ResultRow(
                it.taskKey, it.worker, it.answer,
                accepted = winners[it.taskKey] != null &&
                    normalize(it.answer) == winners[it.taskKey],
            )
        }
    }

    private fun winningAnswer(policy: ValidationPolicy, subs: List<Submitted>): String? {
        val counts = subs.groupingBy { normalize(it.answer) }.eachCount()
        val best = counts.entries
            .sortedWith(compareByDescending<Map.Entry<String, Int>> { it.value }.thenBy { it.key })
            .first()
        val needed = Math.ceil(subs.size * policy.agreementThreshold).toInt().coerceAtLeast(1)
        return if (best.value >= needed) best.key else null
    }

    /**
     * The deterministic payout list: reward per accepted answer, aggregated
     * per worker, ordered by worker pubkey — everyone derives the same list.
     */
    fun payouts(
        rewardPerTaskSats: Long,
        rows: List<ResultRow>,
        payoutAddressOf: (String) -> String,
    ): List<PayoutLine> = rows
        .filter { it.accepted }
        .groupingBy { it.worker }
        .eachCount()
        .toSortedMap()
        .map { (worker, accepted) ->
            PayoutLine(worker, payoutAddressOf(worker), rewardPerTaskSats * accepted)
        }
}

File: ios/HpbCore/Sources/HpbCore/Validators.swift
import Foundation

/// Mechanical validation: pure functions over the revealed submission set.
/// Any observer recomputes both acceptance and the payout list.
public enum Validators {
    /// Kotlin's trim() predicate (Java Character.isWhitespace) — notably it
    /// does NOT trim no-break spaces, unlike Foundation's whitespace set.
    private static func isRefWhitespace(_ scalar: Unicode.Scalar) -> Bool {
        switch scalar.value {
        case 0x09...0x0d, 0x1c...0x1f: return true
        case 0xa0, 0x2007, 0x202f: return false
        default:
            switch scalar.properties.generalCategory {
            case .spaceSeparator, .lineSeparator, .paragraphSeparator: return true
            default: return false
            }
        }
    }

    public static func normalize(_ answer: String) -> String {
        var scalars = Array(answer.unicodeScalars)
        while let first = scalars.first, isRefWhitespace(first) { scalars.removeFirst() }
        while let last = scalars.last, isRefWhitespace(last) { scalars.removeLast() }
        return String(String.UnicodeScalarView(scalars)).lowercased()
    }

    /// The key a tally is grouped by — never the native `String`.
    ///
    /// Swift's `String` equality is canonical equivalence, so `café` written NFC
    /// and NFD compares equal here and collapses to one dictionary key, while
    /// Kotlin's UTF-16 code-unit equality sees two. Two witnesses would compute
    /// different winners and different payouts from identical data. Hex of the
    /// normalized UTF-8 bytes is ASCII, so equality means the same thing on both
    /// sides and depends on neither one's Unicode version.
    public static func labelKey(_ answer: String) -> String {
        Array(normalize(answer).utf8).hex
    }

    /// The commitment format for groundtruth sets: sha256("key:normalized").
    public static func groundtruthHash(_ taskKey: String, _ answer: String) -> String {
        sha256("\(taskKey):\(normalize(answer))").hex
    }

    public struct Submitted {
        public let taskKey: String
        public let worker: String
        public let answer: String

        public init(taskKey: String, worker: String, answer: String) {
            self.taskKey = taskKey
            self.worker = worker
            self.answer = answer
        }
    }

    private static let granted: Set<AssignmentStatus> = [.active, .submitted, .validated]

    /// Grant-scoping, applied by the launcher when collecting AND
    /// re-checked by every witness against the reveal: a row counts only
    /// when its worker holds a granted assignment covering the task, and
    /// only once per (worker, task) — first occurrence wins.
    public static func scoped(_ rows: [Submitted], _ assignments: [AssignmentState]) -> [Submitted] {
        var grantedKeys = [String: Set<String>]()
        for state in assignments where granted.contains(state.status) {
            grantedKeys[state.worker] = Set(state.taskKeys)
        }
        var seen = Set<String>()
        return rows.filter { row in
            grantedKeys[row.worker]?.contains(row.taskKey) == true &&
                seen.insert("\(row.worker)|\(row.taskKey)").inserted
        }
    }

    public static func validate(_ policy: ValidationPolicy, _ submissions: [Submitted]) -> [ResultRow] {
        switch policy.type {
        case .groundtruth:
            return submissions.map {
                ResultRow(
                    taskKey: $0.taskKey, worker: $0.worker, answer: $0.answer,
                    accepted: policy.groundtruthHashes.contains(groundtruthHash($0.taskKey, $0.answer))
                )
            }
        case .agreement:
            return agreement(policy, submissions)
        }
    }

    /// Inter-worker consensus: per task, the modal normalized answer wins
    /// when it reaches ceil(n * threshold); deterministic tie-break by the
    /// lexicographically smallest modal answer.
    private static func agreement(_ policy: ValidationPolicy, _ submissions: [Submitted]) -> [ResultRow] {
        var perTask = [String: [Submitted]]()
        for sub in submissions { perTask[sub.taskKey, default: []].append(sub) }
        let winners = perTask.mapValues { winningAnswer(policy, $0) }
        return submissions.map {
            ResultRow(
                taskKey: $0.taskKey, worker: $0.worker, answer: $0.answer,
                accepted: winners[$0.taskKey].flatMap { $0 } == labelKey($0.answer)
            )
        }
    }

    private static func winningAnswer(_ policy: ValidationPolicy, _ subs: [Submitted]) -> String? {
        var counts = [String: Int]()
        // Tallied by key, tie-broken on the label: hex of UTF-8 does not order
        // the same as UTF-16 above ASCII, and utf16Less is the shared rule.
        var labels = [String: String]()
        for sub in subs {
            let key = labelKey(sub.answer)
            counts[key, default: 0] += 1
            labels[key] = normalize(sub.answer)
        }
        let best = counts.sorted {
            $0.value != $1.value ? $0.value > $1.value
                : utf16Less(labels[$0.key] ?? $0.key, labels[$1.key] ?? $1.key)
        }[0]
        let needed = max(1, Int(ceil(Double(subs.count) * policy.agreementThreshold)))
        return best.value >= needed ? best.key : nil
    }

    /// The deterministic payout list: reward per accepted answer,
    /// aggregated per worker, ordered by worker pubkey.
    public static func payouts(
        _ rewardPerTaskSats: Int64,
        _ rows: [ResultRow],
        _ payoutAddressOf: (String) -> String
    ) -> [PayoutLine] {
        var accepted = [String: Int64]()
        for row in rows where row.accepted { accepted[row.worker, default: 0] += 1 }
        return accepted.keys.sorted(by: utf16Less).map { worker in
            PayoutLine(
                worker: worker,
                address: payoutAddressOf(worker),
                sats: rewardPerTaskSats * accepted[worker]!
            )
        }
    }
}
